src/SharedStream.mo

```diff
@@ -0,0 +1,221 @@
+      histCnt_ := upToId;
+    };
+
+  };
+
+  /// Usage:
+  ///
+  /// let receiver = StreamReceiver<Int>(
+  ///   123
+  ///   func (streamId : Nat, element: Int, index: Nat): () {
+  ///     ... do your logic with incoming item
+  ///   }
+  /// );
+  ///
+  /// Hook-up receive function in the actor class:
+  /// public shared func onStreamChunk(streamId : Nat, chunk: [Int], firstIndex: Nat) : async () {
+  ///   switch (streamId) case (123) { await receiver.onChunk(chunk, firstIndex); }; case (_) { Error.reject("Unknown stream"); }; };
+  /// };
+  class StreamReceiver<T>(
+    streamId : Nat,
+    callback : (streamId : Nat, item : T, index : Nat) -> (),
+  ) {
+
+    var expectedNextIndex_ : Nat = 0;
+
+    public func onChunk(chunk : [T], firstIndex : Nat) : async () {
+      if (firstIndex != expectedNextIndex_) {
+        throw Error.reject("Broken chunk index: " # Nat.toText(firstIndex) # "; expected: " # Nat.toText(expectedNextIndex_));
+      };
+      for (index in chunk.keys()) {
+        callback(streamId, chunk[index], firstIndex + index);
+      };
+      expectedNextIndex_ += chunk.size();
+    };
+
+  };
+
+  /// Usage:
+  ///
+  /// let sender = StreamSender<Int>(
+  ///   123,
+  ///   10,
+  ///   10,
+  ///   func (item) = 1,
+  ///   anotherCanister.appendStream,
+  /// );
+  /// sender.next([1, 2, 3, 4]);
+  /// sender.next([5, 6, 7, 8]);
+  /// sender.next([9, 10, 11, 12]);
+  /// await* sender.sendChunk(); // will send (123, [1..10], 0) to `anotherCanister`
+  /// await* sender.sendChunk(); // will send (123, [11..12], 10) to `anotherCanister`
+  /// await* sender.sendChunk(); // will do nothing, stream clean
+  class StreamSender<T>(
+    streamId : Nat,
+    maxSize : ?Nat,
+    weightLimit : Nat,
+    weightFunc : (item : T) -> Nat,
+    callback : (streamId : Nat, items : [T], firstIndex : Nat) -> async (),
+  ) {
+
+    let queue : TemporaryQueueImpl<T> = TemporaryQueueImpl<T>();
+    // a head of queue before submitting lately failed chunk. Used for error-handling
+    var lowestError : { #Inf; #Val : Nat } = #Inf;
+
+    public func next(items : [T]) : { #ok : Nat; #err : { #NoSpace } } {
+      let finalSize = queue.size() + items.size();
+      switch (maxSize) {
+        case (?max) if (finalSize > max) {
+          return #err(#NoSpace);
+        };
+        case (_) {};
+      };
+      for (item in items.vals()) {
+        queue.push(item);
+      };
+      #ok(finalSize);
+    };
+
+    public func sendChunk() : async* {
+      #ok : Nat;
+      #err : { #Paused; #SubscribtionError : Text };
+    } {
+      switch (lowestError) {
+        case (#Inf) {};
+        case (#Val _) { return #err(#Paused) };
+      };
+      let headId = queue.headId();
+      streamIter.reset();
+      let elements = Iter.toArray(streamIter);
+      if (elements.size() == 0) {
+        return #ok(0);
+      };
+      try {
+        await callback(streamId, elements, headId);
+        queue.pruneHist(headId + elements.size());
```

Comment:
The code after this line from the other comment is still needed here.

Reply:
Done